### PR TITLE
refactor+fix: Medication data validation and Modify persistency

### LINKED ISSUE #170: [PE-D][Tester C] Incorrect dosage in list-t today
Added MedB with dosage 2 in the morning. However, when call list -t today MedB is reflected under morning with dosage 0.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: cayennegoh/ped#5

### LINKED ISSUE #205: [PE-D][Tester B] 'add' command does not handle exceptions where quantity is extremely large
Quantity value being excessively large when adding a medicine causes unhandled exception.

Could implement check in parser to prevent this or could implement exception handler.

-------------
Labels: `type.FunctionalityBug` `severity.Low`
original: LWachtel1/ped#8

### LINKED ISSUE #224: refactor: Validate Medication data before setting
Move all validation to `Medication` setters so all usages of them are validated, be it `add` or `modify`

### PR DESCRIPTION
**Pull request by wenenhoe**
# Checklist for feature implementation and modification
Fixes #205, closes #224, fixes #170
- [x] Added relevant milestone (under milestone tab on the right)
- Implementation
  - [x] Sources (if any) credited under both relevant section(s) and Developer Guide (Acknowledgement)
  - [ ] JUnit Test(s)
- Documentation
  - [x] JavaDoc for classes and methods
  - Developer Guide
    - [ ] Design (Can add as a todo statement for later on)
    - [ ] Implementation (Can add as a todo statement for later on)
    - [ ] Relevant User Story present
    - [ ] Non-functional requirement(s) that the feature addresses (if any)
    - [ ] Glossary of the (technical) term(s) used (if any)
    - [ ] Specialised instructions for manual testing (if any)
  - User Guide
    - [ ] Description and usage of feature (if any)
  - Contributions (under `/docs/team/*.md`)
    - [x] Code/Feature/Enhancement contribution
    - [ ] Developer Guide contribution
    - [ ] User Guide contribution
    - [x] Misc and team-based admin tasks (if any)

# Additional Information

Nil

### CODE REVIEW

File: src/main/java/meditracker/medication/Medication.java [hunk 1]
```diff
@@ -114,30 +151,103 @@ public void setExpiryDate(String expiryDate) {
...
+        case REPEAT:
+            int repeat = convertStringToInteger(argumentValue);
+            setRepeat(repeat);
+            break;
+        case DAY_ADDED:
+            int dayAdded = convertStringToInteger(argumentValue);
+            setDayAdded(dayAdded);
+            break;
+        default:
+            throw new IllegalStateException("Unexpected value: " + argumentName);
```
Comment (annoy-o-mus) [inline src/main/java/meditracker/medication/Medication.java:245]: Will this cause the program to crash if for some reason this gets thrown? It looks like this exception is unhandled. I am thinking of throwing `MediTrackerException` instead.

File: src/main/java/meditracker/medication/Medication.java [hunk 2]
```diff
@@ -239,5 +361,66 @@ public boolean equals(Object o) {
                 && Objects.equals(getRepeat(), medication.getRepeat())
                 && Objects.equals(getDayAdded(), medication.getDayAdded());
     }
+
+    // @@author annoy-o-mus
```
Comment (annoy-o-mus) [inline src/main/java/meditracker/medication/Medication.java:365]: You can have this part of the checks

File: src/main/java/meditracker/medication/Medication.java [hunk 3]
```diff
@@ -200,14 +316,20 @@ public void revertMedication(Medication medication) {
     public static Medication deepCopy(Medication medication) {
         Medication newMedication = new Medication();
 
-        newMedication.setName(medication.getName());
+        try {
+            newMedication.setName(medication.getName());
+            newMedication.setRepeat(medication.getRepeat());
+        } catch (MediTrackerException e) {
+            // critical error as this should not happen
+            throw new RuntimeException(e);
```
Comment (annoy-o-mus) [inline src/main/java/meditracker/medication/Medication.java:324]: Should we try to handle this gracefully instead of crashing?

Comment (annoy-o-mus) [reply src/main/java/meditracker/medication/Medication.java:324]: just like how the remarks have a default "nil" value if null

**Review by annoy-o-mus — CHANGES_REQUESTED** [review]
Great job on the hard work refactoring from many places and addressing the issues brought up by others. Some comments for you to consider.

File: src/test/java/meditracker/storage/JsonExporterTest.java [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
+import meditracker.medication.Medication;
+import meditracker.medication.MedicationManager;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-import meditracker.medication.Medication;
-import meditracker.medication.MedicationManager;
+import java.io.IOException;
```
[the remark this answers is not in the document]
Comment (wenenhoe) [reply src/test/java/meditracker/storage/JsonExporterTest.java:11]: Resolved in 0fc4cf5

File: src/main/java/meditracker/medication/Medication.java [hunk 5]
```diff
@@ -1,11 +1,14 @@
 package meditracker.medication;
 
+import meditracker.argument.ArgumentName;
+import meditracker.exception.MediTrackerException;
 import meditracker.time.Period;
-
-// @@author nickczh
+import org.json.JSONException;
```
[the remark this answers is not in the document]
Comment (wenenhoe) [reply src/main/java/meditracker/medication/Medication.java:5]: Resolved in 0fc4cf5

File: src/main/java/meditracker/medication/Medication.java (hunk 1 above)
Comment (wenenhoe) [reply src/main/java/meditracker/medication/Medication.java]: Resolved in 845916f

File: src/main/java/meditracker/medication/Medication.java (hunk 2 above)
Comment (wenenhoe) [reply src/main/java/meditracker/medication/Medication.java]: Resolved in bc42dd8

File: src/main/java/meditracker/medication/Medication.java (hunk 3 above)
Comment (wenenhoe) [reply src/main/java/meditracker/medication/Medication.java]: Resolved in eb0bf3c

File: src/main/java/meditracker/medication/Medication.java [hunk 6]
```diff
@@ -174,13 +184,23 @@ public int getRepeat() {
         return repeat;
     }
 
+    /**
+     * Sets the repeat value of Medication object without checks
+     *
+     * @param repeat  Repeat value to be set
+     */
+    protected void setRepeat(int repeat) {
+        this.repeat = repeat;
```
Comment (annoy-o-mus) [inline src/main/java/meditracker/medication/Medication.java:193]: Do you want to also subject this to the same 1-7 range check?

**Review by annoy-o-mus — APPROVED** [review]

**Pull request merged**